# 12.1

In [1]:
import pandas as pd 
df=pd.read_csv("3.12.1.csv")
df

y     x1     x2
0    6.40   1.32   1.15
1   15.05   2.69   3.40
2   18.75   3.56   4.10
3   30.25   4.41   8.75
4   44.85   5.35  14.82
5   48.94   6.20  15.15
6   51.55   7.12  15.32
7   61.50   8.87  18.18
8  100.44   9.80  35.19
9  111.42  10.65  40.40

In [2]:
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf
import seaborn as sns
np.seterr(all='print')


results1 = smf.ols('y ~ x1 + x2 ', data=df).fit()
print(results1.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 1.269e+04
Date:                Tue, 30 Jun 2020   Prob (F-statistic):           3.48e-13
Time:                        23:17:26   Log-Likelihood:                -8.2049
No. Observations:                  10   AIC:                             22.41
Df Residuals:                       7   BIC:                             23.32
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.5800      0.607      0.956      0.3

/Users/oswaldo_gomez/Library/Caches/pypoetry/virtualenvs/statistics-4NTnAiTz-py3.8/lib/python3.8/site-packages/scipy/stats/stats.py:1603: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=10
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


Por lo tanto

$$\large{\gamma_{x_1,x2} = \beta_0 + \beta_1x_1 + \beta_2x_2  = 0.58 + 2.71 x_1 + 2.0497 x_2}$$  

# 12.28

In [3]:
df2=pd.read_csv("3.2.12.28.csv")
df2

Y    X1      X2
0  193.0   1.6   851.0
1  172.0  22.0  1058.0
2  113.0  33.0  1357.0
3  230.0  15.5   816.0
4   91.0  43.0  1201.0
5  125.0  40.0  1115.0

In [4]:
results2 = smf.ols('Y ~ X1 + X2 ', data=df2).fit()
print(results2.summary())

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.862
Model:                            OLS   Adj. R-squared:                  0.770
Method:                 Least Squares   F-statistic:                     9.353
Date:                Tue, 30 Jun 2020   Prob (F-statistic):             0.0514
Time:                        23:17:26   Log-Likelihood:                -25.866
No. Observations:                   6   AIC:                             57.73
Df Residuals:                       3   BIC:                             57.11
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    350.9943     74.753      4.695      0.0

/Users/oswaldo_gomez/Library/Caches/pypoetry/virtualenvs/statistics-4NTnAiTz-py3.8/lib/python3.8/site-packages/statsmodels/stats/stattools.py:70: ValueWarning: omni_normtest is not valid with less than 8 observations; 6 samples were given.
  warn("omni_normtest is not valid with less than 8 observations; %i "


In [5]:
df2

Y    X1      X2
0  193.0   1.6   851.0
1  172.0  22.0  1058.0
2  113.0  33.0  1357.0
3  230.0  15.5   816.0
4   91.0  43.0  1201.0
5  125.0  40.0  1115.0

In [6]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from math import sqrt

In [7]:
y = df2['Y']
x = df2[['X1','X2']]

In [8]:
# define the multiple Linear regression model
linear_regress = LinearRegression() 

In [9]:
# Fit the multiple Linear regression model
linear_regress.fit(x,y)
# predict with the data
y_pred = linear_regress.predict(x)
RMSE=sqrt(mean_squared_error(y, y_pred))

In [10]:
2*RMSE**2

650.1407464106162

#### Entonces a)  $\sigma**{2} = 650.14$

Como calculamos anteriormente tenemos que 

$$ {\hat y= 350 -1.27 \beta_1 -0.15 \beta_2  } = 175$$
  
y 

$$  {135 < \mu_{Y|x_1=20,x_2=1000} < 207.4268} $$

y

$$ {82.9 < y_0 < 269.3}$$

# 12.34

In [11]:
df3=pd.read_csv("3.2.12.34.csv")
df3

y  x1  x2  x3   x4
0   240  25  24  91  100
1   236  31  21  90   95
2   290  45  24  88  110
3   274  60  25  87   88
4   301  65  25  91   94
5   316  72  26  94   99
6   300  80  25  87   97
7   296  84  25  86   96
8   267  75  24  88  110
9   276  60  25  91  105
10  288  50  25  90  100
11  261  38  23  89   98

Completo

In [12]:
results3 = smf.ols('y ~ x1 + x2 + x3 + x4', data=df3).fit()
print(results3.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.745
Model:                            OLS   Adj. R-squared:                  0.599
Method:                 Least Squares   F-statistic:                     5.106
Date:                Tue, 30 Jun 2020   Prob (F-statistic):             0.0303
Time:                        23:17:26   Log-Likelihood:                -46.745
No. Observations:                  12   AIC:                             103.5
Df Residuals:                       7   BIC:                             105.9
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   -102.7132    207.859     -0.494      0.6

/Users/oswaldo_gomez/Library/Caches/pypoetry/virtualenvs/statistics-4NTnAiTz-py3.8/lib/python3.8/site-packages/scipy/stats/stats.py:1603: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


# Calculamos RMSE para el modelo completo

In [13]:
y = df3['y']
x = df3[['x1','x2','x3','x4']]
# define the multiple Linear regression model
linear_regress = LinearRegression() 
# Fit the multiple Linear regression model
linear_regress.fit(x,y)
# predict with the data
y_pred = linear_regress.predict(x)
MSE=mean_squared_error(y, y_pred)
RMSE=sqrt(mean_squared_error(y, y_pred))
RMSE

11.898911912581305

# Fórmula para calcular SCR (SSR en inglés) para el modelo completo

In [14]:
X = np.array(x)
y = np.array([y]).T

def get_sum_of_squares():
    ssr = 0
    sse = 0
    sst = 0

    mean_y = np.mean(y)

    for i in range(12):
        y_hat = linear_regress.predict([X[i]])

        ssr += (y_hat - mean_y)**2
        sse += (y[i][0] - y_hat)**2
        sst += (y[i][0] - mean_y)**2

    return ssr[0]
    #return (ssr, sse, sst)

ssr_total = get_sum_of_squares()
ssr_total


4957.240743559574

Aqui podemos hacer uso de la función feature selection que calcula las f's y las p(s) de manera directa. Podemos ver que las p's para  $\beta_1$ y $\beta_2$ son pequeñas por lo que podemos rechazar la hipótesis nula.
Calculemos ahora sin estas dos variables para calcular poco a poco la f reducida

In [15]:
from sklearn.feature_selection import f_regression
#X = df[['Sugars', 'Fiber', 'shelf_1', 'shelf_2']]
#y = df['Rating']
#x = sm.add_constant(x)

(F, pval) = f_regression(x, y)
display(F, pval)

/Users/oswaldo_gomez/Library/Caches/pypoetry/virtualenvs/statistics-4NTnAiTz-py3.8/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


array([1.29737317e+01, 1.81356853e+01, 2.39178819e-02, 8.45968053e-04])

array([0.00483259, 0.00166671, 0.8801709 , 0.97736866])

## Vamos a encontrar ahora un modelo lineal quitando las $\beta_i$ con $i \in \{1,2\}$

In [16]:
results3 = smf.ols('y ~ x3 + x4', data=df3).fit()
print(results3.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.003
Model:                            OLS   Adj. R-squared:                 -0.219
Method:                 Least Squares   F-statistic:                   0.01154
Date:                Tue, 30 Jun 2020   Prob (F-statistic):              0.989
Time:                        23:17:27   Log-Likelihood:                -54.922
No. Observations:                  12   AIC:                             115.8
Df Residuals:                       9   BIC:                             117.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    235.4442    337.987      0.697      0.5

/Users/oswaldo_gomez/Library/Caches/pypoetry/virtualenvs/statistics-4NTnAiTz-py3.8/lib/python3.8/site-packages/scipy/stats/stats.py:1603: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


In [17]:
y = df3['y']
x = df3[['x3','x4']]
# define the multiple Linear regression model
linear_regress = LinearRegression() 
# Fit the multiple Linear regression model
linear_regress.fit(x,y)
# predict with the data
y_pred = linear_regress.predict(x)
MSE=mean_squared_error(y, y_pred)


### Funcion para encontrar SCR reducida


In [18]:
X = np.array(x)
y = np.array([y]).T

def get_sum_of_squares():
    ssr = 0
    sse = 0
    sst = 0

    mean_y = np.mean(y)

    for i in range(12):
        y_hat = linear_regress.predict([X[i]])

        ssr += (y_hat - mean_y)**2
        sse += (y[i][0] - y_hat)**2
        sst += (y[i][0] - mean_y)**2

    return ssr[0]
    #return (ssr, sse, sst)

ssr_reducido = get_sum_of_squares()



In [19]:
((ssr_total-ssr_reducido)/2)/RMSE**2

17.446228770335775

Vemos que la F reducida es del órden de magnitud que habiamos encontrado previamente. Podemos calcular el valor P
    en el órden de las centésimas por lo que rechazamos $H_0$ y podemos decir que $\beta_i \neq 0$ con $i \in \{1,2\}$